PolyEngine/Core/Src/pe/core/storage/Queue.hpp

```diff
@@ -94,182 +89,310 @@ namespace pe::core::storage
+
+			T& operator*() const { return ref.m_data.at(idx); }
+
+			ReverseIterator& operator++() { increment(); return *this; }
+			ReverseIterator operator++(int) { ReverseIterator ret(ref, idx); increment(); return ret; }
+			ReverseIterator& operator--() { idx = ref.getPrevIdx(idx); return *this; }
+			ReverseIterator operator--(int) { ReverseIterator ret(ref, idx); idx = ref.getPrevIdx(idx); return ret; }
+		private:
+			ReverseIterator(Queue<T>& queue, size_t index) : ref(queue), idx(index) {};
+			void increment()
+			{
+				if(idx == ref.m_head)
+					idx = ref.m_tail;
+				else 
+					idx = ref.getPrevIdx(idx);
+			}
+			Queue<T>& ref;
+			size_t idx;
+		};
+
+		class ConstReverseIterator : public BaseObject<>
+		{
+			friend class Queue<T>;
+		public:
+			using iterator_category = std::bidirectional_iterator_tag;
+			using value_type = T;
+			using difference_type = std::ptrdiff_t;
+			using pointer = T*;
+			using reference = T&;
+
+			bool operator==(const ConstReverseIterator& rhs) const { return idx == rhs.idx; }
+			bool operator!=(const ConstReverseIterator& rhs) const { return idx != rhs.idx; }
+
+			const T& operator*() const { return ref.m_data.at(idx); }
+
+			ConstReverseIterator& operator++() { increment(); return *this; }
+			ConstReverseIterator operator++(int) { ConstReverseIterator ret(ref, idx); increment(); return ret; }
+			ConstReverseIterator& operator--() { idx = ref.getPrevIdx(idx); return *this; }
+			ConstReverseIterator operator--(int) { ConstReverseIterator ret(ref, idx); idx = ref.getPrevIdx(idx); return ret; }
+		private:
+			ConstReverseIterator(const Queue<T>& queue, size_t index) : ref(queue), idx(index) {};
+			void increment()
+			{
+				if(idx == ref.m_head)
+					idx = ref.m_tail;
+				else 
+					idx = ref.getPrevIdx(idx);
+			}
+			const Queue<T>& ref;
+			size_t idx;
+		};
+
+		Iterator begin() { return Iterator(*this, m_head); }
+		Iterator end() { return Iterator(*this, m_tail); }
+		ConstIterator cbegin() const { return ConstIterator(*this, m_head); }
+		ConstIterator cend() const { return ConstIterator(*this, m_tail); }
+		ReverseIterator rbegin() { return ReverseIterator(*this, m_tail - 1); }
+		ReverseIterator rend() { return ReverseIterator(*this, m_tail); }
+		ConstReverseIterator crbegin() const { return ConstReverseIterator(*this, m_tail - 1); }
+		ConstReverseIterator crend() const { return ConstReverseIterator(*this, m_tail); }
 
 	private:
 		//------------------------------------------------------------------------------
-		size_t GetPrevIdx(size_t idx) const { return (idx == 0) ? Capacity - 1 : idx - 1; }
-		size_t GetNextIdx(size_t idx) const { return (idx + 1) % Capacity; }
-		void DecreaseIdx(size_t& idx) const { idx = GetPrevIdx(idx); }
-		void AdvanceIdx(size_t& idx) const { idx = GetNextIdx(idx); }
-		size_t GetNthIdx(size_t& n)	const { return (Head + n) % Capacity; }
+		size_t getPrevIdx(size_t idx) const { return (idx == 0) ? m_capacity - 1 : idx - 1; }
```

Comment:
Just thought of optimization, change this line to `return (m_capacity + idx - 1) % m_capacity;`